quota/mysql/mysql_quota.go

```diff
@@ -0,0 +1,149 @@
+// default, even though they are approximate, as they're constant time (select count(*) on InnoDB
+// based MySQL needs to traverse the index and may take quite a while to complete).
+//
+// QuotaManager only implements Global/Write quotas, which is based on the number of Unsequenced
+// rows (to be exact, tokens = MaxUnsequencedRows - actualUnsequencedRows).
+// Other quotas are considered infinite.
+type QuotaManager struct {
+	DB                 *sql.DB
+	MaxUnsequencedRows int
+	UseSelectCount     bool
+}
+
+// GetUser implements quota.Manager.GetUser.
+// User quotas are not implemented by QuotaManager.
+func (m *QuotaManager) GetUser(ctx context.Context, req interface{}) string {
+	return "" // Not used
+}
+
+// GetTokens implements quota.Manager.GetTokens.
+// It doesn't actually reserve or retrieve tokens, instead it allows access based on the number of
+// rows in the Unsequenced table.
+func (m *QuotaManager) GetTokens(ctx context.Context, numTokens int, specs []quota.Spec) error {
+	for _, spec := range specs {
+		if spec.Group != quota.Global || spec.Kind != quota.Write {
+			continue
+		}
+		// Only allow global writes if Unsequenced is under the expected limit
+		count, err := m.countUnsequenced(ctx)
+		if err != nil {
+			return err
+		}
+		if count+numTokens > m.MaxUnsequencedRows {
+			return errors.New("too many unsequenced rows")
+		}
+	}
+	return nil
+}
+
+// PeekTokens implements quota.Manager.PeekTokens.
+// Global/Write tokens reflect the number of rows in the Unsequenced tables, other specs are
+// considered infinite.
+func (m *QuotaManager) PeekTokens(ctx context.Context, specs []quota.Spec) (map[quota.Spec]int, error) {
+	tokens := make(map[quota.Spec]int)
+	for _, spec := range specs {
+		var num int
+		if spec.Group == quota.Global && spec.Kind == quota.Write {
+			count, err := m.countUnsequenced(ctx)
+			if err != nil {
+				return nil, err
+			}
+			num = m.MaxUnsequencedRows - count
+		} else {
+			num = quota.MaxTokens
+		}
+		tokens[spec] = num
+	}
+	return tokens, nil
+}
+
+// PutTokens implements quota.Manager.PutTokens.
+// It's a noop for QuotaManager.
+func (m *QuotaManager) PutTokens(ctx context.Context, numTokens int, specs []quota.Spec) error {
+	return nil
+}
+
+// ResetQuota implements quota.Manager.ResetQuota.
+// It's a noop for QuotaManager.
+func (m *QuotaManager) ResetQuota(ctx context.Context, specs []quota.Spec) error {
+	return nil
+}
+
+func (m *QuotaManager) countUnsequenced(ctx context.Context) (int, error) {
+	if m.UseSelectCount {
+		return countFromTable(ctx, m.DB)
+	}
+	return countFromInformationSchema(ctx, m.DB)
+}
+
+func countFromInformationSchema(ctx context.Context, db *sql.DB) (int, error) {
+	// information_schema.tables doesn't have an explicit PK, so let's play it safe and ensure
+	// the cursor returns a single row.
+	rows, err := db.QueryContext(ctx, countFromInformationSchemaQuery, "Unsequenced", "BASE TABLE")
+	if err != nil {
+		return -1, err
```

Comment:
Done (throughout the PR).

There's no particular meaning to the returned count if `err != nil`.